src/transparency/orig_proto.rs

```diff
@@ -0,0 +1,244 @@
+            // the normal.
+            let was_absolute_form = h1::is_absolute_form(req.uri());
+            if !was_absolute_form {
+                // Since the version is going to set to HTTP_2, the NormalizeUri
+                // middleware won't normalize the URI automatically, so it
+                // needs to be done now.
+                h1::normalize_our_view_of_uri(&mut req);
+            }
+
+            let val = match (req.version(), was_absolute_form) {
+                (http::Version::HTTP_11, false) => "HTTP/1.1",
+                (http::Version::HTTP_11, true) => "HTTP/1.1; absolute-form",
+                (http::Version::HTTP_10, false) => "HTTP/1.0",
+                (http::Version::HTTP_10, true) => "HTTP/1.0; absolute-form",
+                (v, _) => unreachable!("bad orig-proto version: {:?}", v),
+            };
+            req.headers_mut().insert(
+                L5D_ORIG_PROTO,
+                HeaderValue::from_static(val)
+            );
+
+
+            *req.version_mut() = http::Version::HTTP_2;
+            downgrade_response = true;
+        }
+
+        let fut = self.inner.call(req);
+
+        if downgrade_response {
+            fut.map(|mut res| {
+                debug_assert_eq!(res.version(), http::Version::HTTP_2);
+                let version = if let Some(orig_proto) = res.headers().get(L5D_ORIG_PROTO) {
+                    debug!("downgrading {} response: {:?}", L5D_ORIG_PROTO, orig_proto);
+                    if orig_proto == "HTTP/1.1" {
+                        http::Version::HTTP_11
+                    } else if orig_proto == "HTTP/1.0" {
+                        http::Version::HTTP_10
+                    } else {
+                        warn!("unknown {} header value: {:?}", L5D_ORIG_PROTO, orig_proto);
+                        res.version()
+                    }
+                } else {
+                    res.version()
+                };
+                *res.version_mut() = version;
+                res
+            })
+        } else {
+            // Just passing through...
+            fut.map(|res| res)
+        }
+    }
+}
+
+impl<S, B1, B2> NewService for Upgrade<S>
+where
+    S: NewService<Request = http::Request<B1>, Response = http::Response<B2>>,
+{
+    type Request = S::Request;
+    type Response = S::Response;
+    type Error = S::Error;
+    type Service = Upgrade<S::Service>;
+    type InitError = S::InitError;
+    type Future = future::Map<
+        S::Future,
+        fn(S::Service) -> Upgrade<S::Service>
+    >;
+
+    fn new_service(&self) -> Self::Future {
+        let s = self.inner.new_service();
+        // This weird dance is so that the closure doesn't have to
+        // capture `self` and can just be a `fn` (so the `Map`)
+        // can be returned unboxed.
+        if self.upgrade_h1 {
+            s.map(|inner| Upgrade::new(inner, true))
+        } else {
+            s.map(|inner| Upgrade::new(inner, false))
+        }
+    }
```

Comment:
this is equivalent to the following, isn't it?
```rust
  let upgrade_h1 = self.upgrade_h1;
  s.map(|inner| Upgrade::new(inner, upgrade_h1))
}
```

tioli, seems like a slightly less weird dance to just copy the bool imo

Reply:
That way makes the closure capture `upgrade_h1`, and so it can no longer be cast to just a `fn`, and thus is unnameable.